iib/common/tracing.py

```diff
@@ -0,0 +1,210 @@
+
+
+log = logging.getLogger(__name__)
+propagator = TraceContextTextMapPropagator()
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of a calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if TracingWrapper.__instance is None:
+            log.info("Creating TracingWrapper instance")
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint="https://localhost:4317/v1/traces",  # noqa: E501
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: "iib-auto-manual"})
+            )
+            cls.processor = BatchSpanProcessor(otlp_exporter)
+            cls.provider.add_span_processor(cls.processor)
+            trace.set_tracer_provider(cls.provider)
+            set_global_textmap(propagator)
+            cls.tracer = trace.get_tracer(__name__)
+        return cls.__instance
+
+
+def instrument_tracing(
+    func=None,
+    *,
+    service_name: str = "",
+    span_name: str = "",
+    ignoreTracing=False,
+    attributes: Dict = {},
+    existing_tracer: Tracer = None,
+    is_class=False,
+):
+    """
+    Instrument tracing for a function or class.
+
+    :param func_or_class: The function or class to be decorated.
+    :param service_name: The name of the service to be used.
+    :param span_name: The name of the span to be created.
+    :param ignoreTracing: If True, the function will not be traced.
+    :param attributes: The attributes to be added to the span.
+    :param existing_tracer: The tracer to be used.
+    :return: The decorated function or class.
+    """
+
+    def instrument_class(cls):
+        """
+        Instruments class and filters out all the methods of a class that are to be instrumented.
+
+        :param cls: The class to be decorated.
+        :return: The decorated class.
+        """
+        for name, method in cls.__dict__.items():
+            if (
+                callable(method)
+                and not method.__name__.startswith("_")
+                and not inspect.isclass(method)
+            ):
+                setattr(cls, name, instrument_tracing(method))
+        return cls
+
+    def instrument_span(func):
+        log.info(f"Instrumenting span for {span_name}")
+        tracer = trace.get_tracer(__name__)
+        context = None
+        if trace.get_current_span():
+            context = trace.get_current_span().get_span_context()
+        else:
+            context = propagator.extract(carrier={})
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            log.info(f"Context inside {span_name}: {context}")
+            if kwargs.get("traceparent"):
+                log.info(f"traceparent is {kwargs.get('traceparent')}")
```

Comment:
Please use '%s' notation for logging. 

I am sorry, could you please use https://github.com/release-engineering/iib#coding-standards for strings. 
Unfortunately original authors suggested to use single quotes:
- Single quotes are used instead of double quotes

Reply:
It is interesting how this wasn't caught in your black run?

Reply:
F-strings are faster and better. I love this formatting stile - that is why we are using it. However it is not the same for logging. If you use any other formatting for logging then '%s' then the string is formatted every time python hits the line. On the other hand if we use '%s' in logging the formatting is done only if the line will be logged (correct logging level is set). 
In this example if we set logging level to WARNING then f-string will be formatted but the formatted text will never be used. If there is '%s' then nothing happens and the line is skipped.

Reply:
I have updated the tracing.py log statements with %s